services/clearlydefined/clearlydefined-score.service.js

```diff
@@ -0,0 +1,57 @@
+import Joi from 'joi'
+import { nonNegativeInteger } from '../validators.js'
+import { floorCount as floorCountColor } from '../color-formatters.js'
+import { BaseJsonService } from '../index.js'
+
+const schema = Joi.object({
+  scores: Joi.object({
+    effective: nonNegativeInteger,
+  }).required(),
+}).required()
+
+// This service based on the REST API for clearlydefined.io
+// https://api.clearlydefined.io/api-docs/
+export default class ClearlyDefinedService extends BaseJsonService {
+  static category = 'analysis'
+  static route = {
+    base: 'clearlydefined',
+    pattern: 'score/:type/:provider/:namespace/:name/:revision',
+  }
+
+  static examples = [
+    {
+      title: 'ClearlyDefined Score',
+      namedParams: {
+        type: 'npm',
+        provider: 'npmjs',
+        namespace: '-',
+        name: 'jquery',
+        revision: '3.4.1',
+      },
+      staticPreview: this.render({ score: 88 }),
+    },
+  ]
+
+  static defaultBadgeData = { label: 'score' }
+
+  static render({ score }) {
+    score = Math.round(score)
+    return {
+      label: 'score',
+      message: `${score}/100`,
+      color: floorCountColor(score, 40, 60, 100),
+    }
+  }
+
+  async fetch({ type, provider, namespace, name, revision }) {
+    return this._requestJson({
+      schema,
+      url: `https://api.clearlydefined.io/definitions/${type}/${provider}/${namespace}/${name}/${revision}`,
+    })
```

Comment:
in the event of a 5xx status code the response handling framework will set the badge message to an according error (it also does so with other codes like 404s, but not necessary here). if you want to map a status code to a specific error that can be done via the `errorMessages` object passed in the options to _requestJson and friends, e.g.

```suggestion
    return this._requestJson({
      schema,
      url: `https://api.clearlydefined.io/definitions/${type}/${provider}/${namespace}/${name}/${revision}`,
      errorMessages: {
        500: 'unknown type, provider, or upstream issue',
      },
    })
```

i think the actual error text could be a bit better so don't feel obligated to use that literally, we just want to account for the fact that we could see a legitimate 500 error response even with a valid type and valid provider

Reply:
The actual error is "An error has occurred" so your suggestion is better :-)